packages/ui/src/utils/react.ts

```diff
@@ -0,0 +1,25 @@
+import { ComponentPropsWithRef, ElementType } from 'react';
+
+/**
+ * Utility to make it easier to type props when you're forwarding most props
+ * to the inner component, e.g. forwarding Tamagui style props.
+ *
+ * ```tsx
+ * function MyButton({
+ *   children,
+ *   name,
+ *   ...forwardedProps
+ * }: ForwardingProps<
+ *   typeof Button, // the inner component type
+ *   { myProp: string }, // custom props that should be included in props
+ *   'disabled' | 'onClick'  // which inner props should *not* be exposed?
+ * >) {
+ *   return <Button {...forwardedProps}> ... </Button>;
+ * }
+ * ```
+ */
+export type ForwardingProps<
+  E extends ElementType,
+  CustomProps extends Record<string, unknown>,
+  OmitKeys extends keyof ComponentPropsWithRef<E> = never,
+> = CustomProps & Omit<ComponentPropsWithRef<E>, keyof CustomProps | OmitKeys>;
```

Comment:
I was doing this a couple of times so decided to make it a shared thing.

I use [this wrapper](https://gist.github.com/davidisaaclee/fdd1baa1d5fed4463923ea27bfc04522) in my personal projects to do similar stuff with stronger type guarantees (ensures you don't expose "connected" props; automatically handles refs), which would look like:

```tsx
export const DetailPostView = connectComponent(
  typeof YStack,
  function ({
    children: BaseComponent,
    ownProps: { post, navigateBack },
  }: ConnectorProps<
    { post: db.Post; navigateBack?: () => void },
    PartialProps<typeof YStack, "backgroundColor" | "children">
  >) {
    return (
      <BaseComponent backgroundColor="...">
        <ImplementationBlahBlah post={post} onPress={navigateBack} />
      </BaseComponent>
    );
  }
);
```

`ForwardingProps` approach is definitely simpler!

Reply:
Nice! This is helpful. In some cases tamagui's styleable can be useful for this too, but only works for tamagui components, and may have a little overhead. Looks like: 

```ts
export const TextInputWithIcon = React.memo(
  BaseTextInput.styleable<{ icon: IconType }>(({ icon, ...props }, ref) => {
    return (
      <XStack
        borderRadius="$l"
        borderWidth={1}
        borderColor="$border"
        alignItems="center"
        paddingLeft="$xl"
        gap="$l"
      >
        <Icon type={icon} customSize={['$2xl', '$2xl']} />
        <BaseTextInput
          paddingLeft={0}
          borderWidth={0}
          borderRadius={0}
          flex={1}
          ref={ref}
          {...props}
        />
      </XStack>
    );
  })
);
```

Reply:
oh yeah – I was using that for `Carousel` but iirc, I couldn't figure out how to add a custom ref type...